crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,113 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    time::Instant,
+};
+
+use parking_lot::RwLock;
+use telio_crypto::PublicKey;
+use telio_model::api_config::EndpointProvider;
+use telio_utils::telio_log_warn;
+
+// Possible connection types - relayed ones allow adding info about the other endpoint
+#[derive(Copy, Clone, Debug)]
+pub enum ConnectionType {
+    Relayed,
+    Local(Option<EndpointProvider>),
+    UPnP(Option<EndpointProvider>),
+    UDPHolePunch(Option<EndpointProvider>),
+}
+
+// Possible disconnection reasons
+#[derive(Copy, Clone, Debug)]
+pub enum DisconnectionReason {
+    Initial,
+    DisabledByUser,
+    NetworkError,
+}
+
+// Possible connectivity state changes
+#[derive(Copy, Clone, Debug)]
+pub enum ConnectivityInfo {
+    Disconnected(DisconnectionReason),
+    Connecting,
+    Connected(ConnectionType),
+}
+
+// Event containing info about changed connection quality to a certain peer
+#[derive(Copy, Clone, Debug)]
+pub struct ConnectivityChangeEvent {
+    timestamp: Instant,
+    peer: PublicKey,
+    info: ConnectivityInfo,
+}
+
+// Trait representing generic aggregator of the connectivity data
+pub trait ConnectivityDataAggregatorTrait {
+    // Interface function for data collectors like telio-relay and telio-traversal
+    fn add_state_change(&mut self, event: ConnectivityChangeEvent);
+}
+
+// Trait representing generic connectivity data storage
+trait ConnectivityDataStorageTrait {
+    // Get the new info about connectivity changes
+    fn get_unacknowledged_states(&self) -> Vec<ConnectivityChangeEvent>;
+
+    // Acknowledge which info can be already discarded
+    fn acknowledge_state_transmission<T>(&mut self, info_key_set: T)
+    where
+        T: IntoIterator<Item = (Instant, PublicKey)>;
+}
+
+// A container to store the connectivity data for our meshnet peers
+struct ConnectivityDataAggregator {
+    // TODO: decide which data structure is the best choice
+    storage: RwLock<BTreeMap<(Instant, PublicKey), ConnectivityInfo>>,
+}
+
+impl ConnectivityDataAggregator {
+    // Create a new DataConnectivityAggregator instance
+    fn new() -> ConnectivityDataAggregator {
+        ConnectivityDataAggregator {
+            storage: parking_lot::RwLock::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl ConnectivityDataAggregatorTrait for ConnectivityDataAggregator {
+    fn add_state_change(&mut self, event: ConnectivityChangeEvent) {
+        let mut storage_guard = self.storage.write();
+
+        if let Entry::Vacant(e) = storage_guard.entry((event.timestamp, event.peer)) {
+            e.insert(event.info);
+        } else {
+            telio_log_warn!("Discarding connectivity info - duplicate connectivity event");
+        }
+    }
+}
+
+impl ConnectivityDataStorageTrait for ConnectivityDataAggregator {
+    fn get_unacknowledged_states(&self) -> Vec<ConnectivityChangeEvent> {
```

Comment:
IMHO, in my opinion, it would be nice to handle the "unfinished states" in the aggregator. What I mean by that is:
* `telio-traversal` and `telio-relay` sends raw "started/finished" events
* Aggregator gets them and stores them
* Whenever heartbeat occurs, aggregator finds all of the events, converts them from two separate events (start/finish) into a span of the state (e.g. what state, how long it continued, endpoint provider types and/or reasons for state changes etc.)
* If there is a start event, but no finish event for longer then 24h - aggregator converts it into a "finished" state of 24h and new one started.